shakedown/dcos/spinner.py

```diff
@@ -1,36 +1,49 @@
 import time as time_module
+import traceback
 
+import shakedown
 
-def wait_for(predicate, timeout_seconds=120, sleep_seconds=1, ignore_exceptions=True, inverse_predicate=False):
-    """ waits or spins for a predicate.  Predicate is in function that returns a True or False.
+
+def wait_for(predicate, timeout_seconds=120, sleep_seconds=1, ignore_exceptions=True, inverse_predicate=False, noisy=True):
+    """ waits or spins for a predicate, returning the result.
+        Predicate is a function that returns a truthy or falsy value.
         An exception in the function will be returned.
         A timeout will throw a TimeoutExpired Exception.
 
     """
 
+    start_time = time_module.time()
     timeout = Deadline.create_deadline(timeout_seconds)
     while True:
         try:
             result = predicate()
         except Exception as e:
-            if not ignore_exceptions:
-                raise e
+            if ignore_exceptions:
+                if noisy:
+                    traceback.print_exc()
+            else:
+                raise # preserve original stack
         else:
             if (not inverse_predicate and result) or (inverse_predicate and not result):
-                return
+                return result
             if timeout.is_expired():
                 raise TimeoutExpired(timeout_seconds, str(predicate))
```

Comment:
while you are here:)  if you could... the `str(predicate)` doesn't return useful information... lambda blah...  it isn't possible to see on timeout what we were waiting for... could you update to `str(predicate.__name__)` or something useful.